### PR TITLE
CORE-1904 StreamMessageValidator abstraction

### PR DESCRIPTION
**Pull request by hpihkala**
Adds a shared piece of code, `StreamMessageValidator`, which implements the validity checks of all `StreamMessage`s the Streamr protocol.

Currently similar checks are implemented in both `broker` and `streamr-client-javascript` separately. The goal of this class is to enable refactoring out the overlapping parts in both of the above repos to make the combination of codebases more DRY.

This change is the first subtask in implementing the [validation logic for group key requests and responses](https://streamr.atlassian.net/browse/CORE-1883).

Also added a caching wrapper `CachingStreamMessageValidator` to remove the need to repeat caching logic in `broker` and `streamr-client-javascript`.

### CODE REVIEW

File: src/utils/StreamMessageValidator.js [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    constructor(getStreamFn, isPublisherFn, isSubscriberFn, recoverAddressFn) {
+        StreamMessageValidator.checkInjectedFunctions(getStreamFn, isPublisherFn, isSubscriberFn, recoverAddressFn)
+        this.getStream = getStreamFn
+        this.isPublisher = isPublisherFn
+        this.isSubscriber = isSubscriberFn
+        this.recoverAddress = recoverAddressFn
+    }
+
+    static checkInjectedFunctions(getStreamFn, isPublisherFn, isSubscriberFn, recoverAddressFn) {
+        if (!getStreamFn) {
```
[the remark this answers is not in the document]
Comment (timoxley) [reply src/utils/StreamMessageValidator.js:35]: Same for the other checks below

**Comment by timoxley** [comment]
> Also added a caching wrapper CachingStreamMessageValidator to remove the need to repeat caching logic in broker and streamr-client-javascript

👍 👍

File: src/index.js [hunk 2]
```diff
@@ -132,11 +134,14 @@ export const Errors = {
     InvalidJsonError,
     UnsupportedVersionError,
     GapFillFailedError,
+    ValidationError,
```
Comment (timoxley) [inline src/index.js:137]: `ValidationError` is not defined

File: test/unit/utils/StreamMessageValidator.test.js [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+        groupKeyResponse = StreamMessageFactory.deserialize('[31,["SYSTEM/keyexchange/0xbce3217F2AC9c8a2D14A6303F87506c4FC124014",0,1587143432683,0,"0xFeAACDBBc318EbBF9BB5835D4173C1a7fC24B3b9","2hmxXpkhmaLcJipCDVDm"],null,29,1,"{\\"streamId\\":\\"tagHE6nTQ9SJV2wPoCxBFw\\",\\"keys\\":[{\\"groupKey\\":\\"encrypted-group-key\\",\\"start\\":34524}]}",2,"0xab2ecce6ee5cfb0890cbb4f9f4d5daf9fd8283eb53bdf67144d920624e22424657f96c3c2a3cafe53ca2049e235be5d08f6181546152c3e61cf509db2fd0e6701c"]')
+        groupKeyReset = StreamMessageFactory.deserialize('[31,["SYSTEM/keyexchange/0xbce3217F2AC9c8a2D14A6303F87506c4FC124014",0,1587143432683,0,"0xFeAACDBBc318EbBF9BB5835D4173C1a7fC24B3b9","2hmxXpkhmaLcJipCDVDm"],null,30,1,"{\\"streamId\\":\\"tagHE6nTQ9SJV2wPoCxBFw\\",\\"groupKey\\":\\"encrypted-group-key\\",\\"start\\":34524}",2,"0xefa15b14db0f57c8c6d14fd8428d8de195db691598a53000dc2993808563edef49118cb666581b260f0fe70874f52553b2851dcb19b5f19333136191211cff2b1b"]')
```
[the remark this answers is not in the document]
Comment (hpihkala) [reply test/unit/utils/StreamMessageValidator.test.js:40]: The protocol is versioned, so these literals are immutable and should work until support for this version (`StreamMessage` version `31`) is dropped. That's the latest version, so it won't happen soon.

The message validation is likely to change with new protocol versions, so the test messages will probably be updated whenever the validation logic changes in `StreamMessageValidator`.

**Comment by timoxley** [comment]
@hpihkala I'll fix these conflicts and merge

**Pull request merged**